src/index.ts

```diff
@@ -0,0 +1,216 @@
+ */
+
+function policyToString(directives: CSPDirectives, reportUri?: string): string {
+	let cspStringParts: string[] = [];
+
+	for (let directiveName in directives) {
+		if (!directives.hasOwnProperty(directiveName)) {
+			continue;
+		}
+
+		let directiveValue: CSPDirectiveValue = directives[directiveName as keyof CSPDirectives];
+		let directiveRulesString = getDirectiveString(
+			directiveName as CSPDirectiveName,
+			directiveValue
+		);
+
+		if (directiveRulesString) {
+			cspStringParts.push(directiveRulesString);
+		}
+	}
+
+	if (reportUri) {
+		cspStringParts.push(getReportUriDirective(reportUri));
+	}
+
+	return cspStringParts.join(' ');
+}
+
+/**
+ * Build directive rules part of CSP header value
+ */
+function getDirectiveString(directiveName: CSPDirectiveName, directiveValue: CSPDirectiveValue): string {
+	if (!directiveValue) {
+		return '';
+	}
+
+	if (directiveName in BOOLEAN_DIRECTIVES) {
+		return `${directiveName};`;
+	}
+
+	if (directiveName in STRING_DIRECTIVES) {
+		return `${directiveName} ${directiveValue};`;
+	}
+
+	if (directiveName in LIST_DIRECTIVES) {
+		let valueString = (directiveValue as CSPListDirectiveValue).join(', ');
+		return `${directiveName} ${valueString};`;
+	}
+}
+
+/**
+ * Build report-uri directive
+ */
+function getReportUriDirective(reportUri: string): string {
+	return `report-uri ${reportUri};`;
+}
+
+/**
+ * Normalize different presets list formats to array format
+ */
+function normalizePresetsList(presets: CSPPreset): CSPPresetsArray {
+	return Array.isArray(presets) ? presets : Object.values(presets);
+}
+
+/**
+ * Merges presets to policy
+ */
+function applyPresets(directives: CSPDirectives, presets: CSPPresetsArray): CSPDirectives {
+	let mergedPolicies: CSPDirectives = {};
+
+	for (let preset of [directives, ...presets]) {
+		for (let directiveName in preset) {
+			if (!(directiveName in ALLOWED_DIRECTIVES)) {
+				continue;
+			}
+
+			directiveName as keyof CSPDirectives;
+
+			let currentRules: CSPDirectiveValue = mergedPolicies[directiveName as keyof CSPDirectives];
+			let presetRules: CSPDirectiveValue = preset[directiveName as keyof CSPDirectives];
+
+			(mergedPolicies[directiveName as keyof CSPDirectives] as CSPDirectiveValue) = mergeDirectiveRules(currentRules, presetRules, directiveName as keyof CSPDirectives);
+		}
+	}
+
+	return mergedPolicies;
+}
+
+function mergeDirectiveRules(directiveValue1: CSPDirectiveValue, directiveValue2: CSPDirectiveValue, directiveName: CSPDirectiveName): CSPDirectiveValue {
+	if (directiveValue1 === undefined) {
+		return directiveValue2;
+	}
+
+	if (directiveValue2 === undefined) {
+		return directiveValue1;
+	}
+
+	if (directiveName in LIST_DIRECTIVES) {
+		return getUniqRules([
+			...directiveValue1 as CSPListDirectiveValue,
+			...directiveValue2 as CSPListDirectiveValue
+		]);
+	}
+
+	return directiveValue2;
+}
+
+function getUniqRules(rules: CSPListDirectiveValue): CSPListDirectiveValue {
+	return Array.from(new Set(rules));
```

Comment:
I suppose garbage collector will work fine here, but maybe WeakSet?

Reply:
It would work if rules was an array of objects. But it's an array of string values.